### PR TITLE
Changing pattoken input to service endpoint type

### PR DESCRIPTION
**Pull request by tejasd1990**

### CODE REVIEW

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 1]
```diff
@@ -199,12 +200,11 @@
             "helpMarkDown": "Choose if the tags configured on the Azure VM need to be copied to the corresponding machine group agent. <br><br>​By default all Azure tags will be copied following the format “Key: Value”. Example: An Azure Tag “Role : Web” would be copied  as-is to the Agent machine. <br><br>For more information on how tag Azure resources refer to [link](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-using-tags​)"
         },
         {
-            "name": "vstsPATToken",
-            "type": "string",
+            "name": "vstsPATTokenEndpoint",
+            "type": "connectedService:ExternalTfs",
             "label": "Personal access token",
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/task.json]: label should change. talk with PM.

File: Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson [hunk 2]
```diff
@@ -33,8 +33,8 @@
   "loc.input.help.machineGroupName": "Select a machine group",
   "loc.input.label.copyAzureVMTags": "Copy Azure VM tags to agents",
   "loc.input.help.copyAzureVMTags": "Choose if the tags configured on the Azure VM need to be copied to the corresponding machine group agent. <br><br>​By default all Azure tags will be copied following the format “Key: Value”. Example: An Azure Tag “Role : Web” would be copied  as-is to the Agent machine. <br><br>For more information on how tag Azure resources refer to [link](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-using-tags​)",
-  "loc.input.label.vstsPATToken": "Personal access token",
-  "loc.input.help.vstsPATToken": "Agent registration with machine group requires access to your visual studio project.​ It is recommended that you store the token in a secure variable and provide the variable reference here.",
+  "loc.input.label.vstsPATTokenEndpoint": "Personal access token",
+  "loc.input.help.vstsPATTokenEndpoint": "Agent registration with machine group requires access to your visual studio project.​ It is recommended that you store the token in a secure variable and provide the variable reference here.",
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson]: I guess these strings should change to reflect service endpoints

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 3]
```diff
@@ -199,12 +200,11 @@
...
             "groupName": "AdvancedDeploymentOptions",
             "visibleRule": "enableDeploymentPrerequisites = ConfigureVMWithMGAgent",
-            "defaultValue": "",
             "required": true,
             "helpMarkDown": "Agent registration with machine group requires access to your visual studio project.​ It is recommended that you store the token in a secure variable and provide the variable reference here."
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/task.json]: this will also change

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 4]
```diff
@@ -230,13 +230,15 @@
...
-    ],
-    "sourceDefinitions": [
+        ,
         {
             "target": "machineGroupName",
-            "endpoint": "/$(system.teamProject)/_apis/distributedtask/machinegroups",
-            "selector": "jsonpath:$.value[*].name",
-            "authKey": "tfs:teamfoundation"
+            "endpointId": "$(vstsPATTokenEndpoint)",
+            "dataSourceName": "MachineGroups",
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/task.json]: DeploymentGroups. I think this you have already done.

File: Tasks/AzureResourceGroupDeployment/Tests/addVSTSExtension.ts [hunk 5]
```diff
@@ -17,17 +17,18 @@ tr.setInput("csmParametersFile", path.join(__dirname, process.env["csmParameters
 tr.setInput("enableDeploymentPrerequisites", process.env["enableDeploymentPrerequisites"]);
 tr.setInput("machineGroupName", "biprasad");
 tr.setInput("copyAzureVMTags", process.env["copyAzureVMTags"]);
-tr.setInput("vstsPATToken", "PAT");
+tr.setInput("vstsPATTokenEndpoint", "PatEndpoint");
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/Tests/addVSTSExtension.ts]: Do you have a test case which tests a wrong PAT token?

Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/Tests/addVSTSExtension.ts]: Also add a test case which tests service endpoints with non-token auth

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 6]
```diff
@@ -45,12 +70,13 @@ export class AzureRGTaskParameters {
             this.enableDeploymentPrerequisites = tl.getInput("enableDeploymentPrerequisites");
             this.machineGroupName = tl.getInput("machineGroupName");
             this.copyAzureVMTags = tl.getBoolInput("copyAzureVMTags");
-            this.vstsPATToken = tl.getInput("vstsPATToken");
+            var vstsPatTokenEndpointName = tl.getInput("vstsPATTokenEndpoint", false);
+            if(vstsPatTokenEndpointName){
+                this.tokenCredentials = this.getVSTSPatToken(vstsPatTokenEndpointName);
+            }
             this.outputVariable = tl.getInput("outputVariable");
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: line space please

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 7]
```diff
@@ -59,6 +85,19 @@ export class AzureRGTaskParameters {
         }
     }
 
+    private getVSTSPatToken(vstsPATTokenEndpointName: string): TokenCredentials{
+        var endpointAuth = tl.getEndpointAuthorization(vstsPATTokenEndpointName, true);
+        if(endpointAuth.scheme == 'Token'){
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: ===

File: Tasks/AzureResourceGroupDeployment/operations/MachineGroupExtensionHelper.ts [hunk 8]
```diff
@@ -192,7 +192,7 @@ export class MachineGroupExtensionHelper {
             var autoUpgradeMinorVersion: boolean = true;
             var publisher: string = this.publisher;
             var extensionType: string = this.extensionType;
-            var collectionUri = this.taskParameters.machineGroupCollectionUrl;
+            var collectionUri = this.taskParameters.tokenCredentials.getHostUrl();
             var teamProject = this.taskParameters.machineGroupProjectName;
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/operations/MachineGroupExtensionHelper.ts]: as discussed, team project will come from task input which will populated from endpoint data source.

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 9]
```diff
@@ -4,6 +4,32 @@
 import tl = require("vsts-task-lib/task");
 import msRestAzure = require("./../operations/azure-rest/azure-arm-common");
 
+class TokenCredentials {
+    private hostUrl: string;
+    private patToken: string;
+
+    constructor(hostUrl: string, patToken: string){
+        if (!Boolean(hostUrl) || typeof hostUrl.valueOf() !== 'string') {
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: these two checks must be done before calling this constructor as well. And error message should clearly say that endpoint's url/token field must not be empty.

File: Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson [hunk 10]
```diff
@@ -160,5 +160,7 @@
   "loc.messages.TimeoutWhileWaiting": "Timed out while waiting",
   "loc.messages.InvalidTemplateLocation": "The template location supplied is invalid. Task only supports 'Linked artifact' or 'URL of the file'",
   "loc.messages.EncodingNotSupported": "Encoding of the file '%s' is '%s' which is not supported. Supported encodings are ['utf-8', 'utf-16le']",
-  "loc.messages.CouldNotDetectEncoding": "Could not detect encoding of file '%s'"
+  "loc.messages.CouldNotDetectEncoding": "Could not detect encoding of file '%s'",
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson]: where is this used?

**Comment by bryanmacfarlane** [comment]
Is this a compat breaking change?  What happens to users that were using the field you set?

**Comment by tejasd1990** [comment]
The pattoken field which is being changed, is part of the 'configuration against machine group' option. This option has not yet been enabled in UI, so there are no existing customers using it.

File: Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson (hunk 10 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/Strings/resources.resjson/en-US/resources.resjson]: This is already existing part.

File: Tasks/AzureResourceGroupDeployment/task.json (hunk 4 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/task.json]: Yes

File: Tasks/AzureResourceGroupDeployment/operations/MachineGroupExtensionHelper.ts (hunk 8 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/operations/MachineGroupExtensionHelper.ts]: Done

File: Tasks/AzureResourceGroupDeployment/Tests/addVSTSExtension.ts (hunk 5 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/Tests/addVSTSExtension.ts]: Added both

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 9 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: Added

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 11]
```diff
@@ -351,24 +369,29 @@
...
         "VMStartFailed": "Could not start the virtual machine %s. Error : %s",
         "VMStarted": "Started virtual machine: %s",
         "VMTransitioningSkipExtensionAddition": "Virtual machine %s is in a transitioning state. Skipping adding extension on it.",
         "VMDetailsFetchFailed": "Could not get details of virtual machine %s. Error : %s",
         "VMDetailsFetchSucceeded": "Fetched details of virtual machine %s.",
-        "AddingMGAgentOnVMs": "Adding machine group agent on the resource group virutal machines.",
-        "DeletingMGAgentFromVMs": "Deleting machine group agent from the resource group virutal machines.",
-        "DeletingMGAgentOnVMs": "Deleting machine group agent on virtual machines",
+        "AddingDGAgentOnVMs": "Adding deployment group agent on the resource group virutal machines.",
+        "DeletingDGAgentFromVMs": "Deleting deployment group agent from the resource group virutal machines.",
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/task.json]: Typo in virtual

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 12]
```diff
@@ -351,24 +369,29 @@
...
+        "DeletingDGAgentOnVMs": "Deleting deployment group agent on virtual machines",
         "AddingExtensionFailed": "Addition of extension on vm %s failed",
         "TimeoutWhileWaiting": "Timed out while waiting",
         "InvalidTemplateLocation": "The template location supplied is invalid. Task only supports 'Linked artifact' or 'URL of the file'",
         "EncodingNotSupported": "Encoding of the file '%s' is '%s' which is not supported. Supported encodings are ['utf-8', 'utf-16le']",
-        "CouldNotDetectEncoding": "Could not detect encoding of file '%s'"
+        "CouldNotDetectEncoding": "Could not detect encoding of file '%s'",
+        "HostUrlCannotBeEmpty": "Host url must be a non empty string.",
+        "PatTokenCannotBeEmpty": "Pat token must be a non empty string.",
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/task.json]: Lets use full form Personal Access Token

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 9 above)
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: typeof hostUrl.valueOf() !== 'string' || !hostUrl

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 13]
```diff
@@ -43,22 +68,44 @@ export class AzureRGTaskParameters {
...
+
+            if (!Boolean(patToken) || typeof patToken.valueOf() !== 'string') {
+                throw new Error(tl.loc("DeploymentGroupEndpointPatTokenCannotBeEmpty"));
+            }
+            var credentials = new TokenCredentials(hostUrl, patToken);
+            return credentials;
+        }
+        else {
+            var msg = tl.loc("OnlyTokenAuthAllowed");
+            console.log(msg);
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: Do you need to do console.log() as well

Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: For L0 test.

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 9 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: Done

File: Tasks/AzureResourceGroupDeployment/task.json (hunk 12 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/task.json]: Done

File: Tasks/AzureResourceGroupDeployment/task.json (hunk 11 above)
Comment (tejasd1990) [inline Tasks/AzureResourceGroupDeployment/task.json]: Done

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 13 above)
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: you don't need to do console.log for that. TestFx will print loc_OnlyTokenAuthAllowed which can be used for validation.

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 14]
```diff
@@ -43,22 +68,44 @@ export class AzureRGTaskParameters {
...
             throw new Error(tl.loc("ARGD_ConstructorFailed", error.message));
         }
     }
 
+    private getVSTSPatToken(deploymentGroupEndpointName: string): TokenCredentials {
+        var endpointAuth = tl.getEndpointAuthorization(deploymentGroupEndpointName, true);
+        if (endpointAuth.scheme === 'Token') {
+            var hostUrl = tl.getEndpointUrl(deploymentGroupEndpointName, true);
+            var patToken: string = endpointAuth.parameters["apitoken"];
+            if (!Boolean(hostUrl) || typeof hostUrl.valueOf() !== 'string') {
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: just check for null. check type only if endpointAuth.Parameters can be anything other than a string as well

File: Tasks/AzureResourceGroupDeployment/task.json [hunk 15]
```diff
@@ -175,17 +175,38 @@
...
+            "visibleRule": "enableDeploymentPrerequisites = ConfigureVMWithDGAgent",
+            "required": true,
+            "helpMarkDown": "Agent registration with Deployment group requires access to your Visual Studio project.​ <br><br>Click \"Add\" to create an endpoint using personal access token (PAT) with scope restricted to \"Deployment Group\" and a default expiration time of 90 days. <br><br>​Click \"Manage\" to update endpoint details.​"
+        },
+        {
+            "name": "project",
             "type": "pickList",
-            "label": "Machine group",
+            "label": "Team project",
+            "defaultValue": "",
```
Comment (bishal-pdMSFT) [inline Tasks/AzureResourceGroupDeployment/task.json]: remove default value

**Pull request merged**